src/Pact/Types/Term/Internal.hs

```diff
@@ -0,0 +1,1047 @@
+-- ModuleGuard
+
+data ModuleGuard = ModuleGuard
+  { _mgModuleName :: !ModuleName
+  , _mgName :: !Text
+  } deriving (Eq,Generic,Show,Ord)
+
+instance NFData ModuleGuard
+
+instance Pretty ModuleGuard where
+  pretty ModuleGuard{..} = "ModuleGuard" <+> commaBraces
+    [ "module: " <> pretty _mgModuleName
+    , "name: " <> pretty _mgName
+    ]
+
+instance SizeOf ModuleGuard where
+  sizeOf (ModuleGuard md n) =
+    (constructorCost 2) + (sizeOf md) + (sizeOf n)
+
+moduleGuardProperties :: JsonProperties ModuleGuard
+moduleGuardProperties o =
+  [ "moduleName" .= _mgModuleName o
+  , "name" .= _mgName o
+  ]
+{-# INLINEABLE moduleGuardProperties #-}
+
+instance ToJSON ModuleGuard where
+  toJSON = enableToJSON "Pact.Types.Term.ModuleGuard" . lensyToJSON 3
+  toEncoding = A.pairs . mconcat . moduleGuardProperties
+  {-# INLINEABLE toJSON #-}
+  {-# INLINEABLE toEncoding #-}
+
+instance FromJSON ModuleGuard where parseJSON = lensyParseJSON 3
+
+-- -------------------------------------------------------------------------- --
+-- PactGuard
+
+data PactGuard = PactGuard
+  { _pgPactId :: !PactId
+  , _pgName :: !Text
+  } deriving (Eq,Generic,Show,Ord)
+
+instance NFData PactGuard
+
+instance Pretty PactGuard where
+  pretty PactGuard{..} = "PactGuard" <+> commaBraces
+    [ "pactId: " <> pretty _pgPactId
+    , "name: "   <> pretty _pgName
+    ]
+
+instance SizeOf PactGuard where
+  sizeOf (PactGuard pid pn) =
+    (constructorCost 2) + (sizeOf pid) + (sizeOf pn)
+
+pactGuardProperties :: JsonProperties PactGuard
+pactGuardProperties o =
+  [ "pactId" .= _pgPactId o
+  , "name" .= _pgName o
+  ]
+{-# INLINE pactGuardProperties #-}
+
+instance ToJSON PactGuard where
+  toJSON = enableToJSON "Pact.Types.Term.PactGuard" . lensyToJSON 3
+  toEncoding = A.pairs . mconcat . pactGuardProperties
+  {-# INLINEABLE toJSON #-}
+  {-# INLINEABLE toEncoding #-}
+
+instance FromJSON PactGuard where parseJSON = lensyParseJSON 3
+
+-- -------------------------------------------------------------------------- --
+-- ObjectMap
+
+-- | Simple dictionary for object values.
+newtype ObjectMap v = ObjectMap { _objectMap :: M.Map FieldKey v }
+  deriving (Eq,Ord,Show,Functor,Foldable,Traversable,Generic,SizeOf)
+
+instance NFData v => NFData (ObjectMap v)
+
+-- | O(n) conversion to list. Adapted from 'M.toAscList'
+objectMapToListWith :: (FieldKey -> v -> r) -> ObjectMap v -> [r]
+objectMapToListWith f (ObjectMap m) = M.foldrWithKey (\k x xs -> (f k x):xs) [] m
+
+instance Pretty v => Pretty (ObjectMap v) where
+  pretty om = annotate Val $ commaBraces $
+    objectMapToListWith (\k v -> pretty k <> ": " <> pretty v) om
+
+-- FIXME the order depends on the hash function that is used by HM.HashMap in
+-- aeson.
+--
+instance ToJSON v => ToJSON (ObjectMap v) where
+  toJSON (ObjectMap om) = toJSON $ M.mapKeys asString om
+  toEncoding (ObjectMap om) = toEncoding
+    $ HM.fromList
+    $ first (LegacyHashed . asString)
```

Comment:
this is probably ok to NOT rename, as it clearly repros a legacy case ... so maybe ok to keep legacyHashMap et al.